### PR TITLE
build: upgrade to use nodev20

### LINKED ISSUE #11404: `digital envelope routines::unsupported` issue with 3.4.9 release build
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

```
node:internal/crypto/hash:69
    this[kHandle] = new _Hash(algorithm, xofLen);
                    ^
  
  Error: error:0308010C:digital envelope routines::unsupported
      at new Hash (node:internal/crypto/hash:69:19)
      at Object.createHash (node:crypto:138:10)
      at module.exports (/private/tmp/argo-20230720-48[394](https://github.com/Homebrew/homebrew-core/actions/runs/5612007335/jobs/10269330959?pr=137118#step:3:395)-1xggx70/ui/node_modules/webpack/lib/util/createHash.js:135:53)
      at NormalModule._initBuildHash (/private/tmp/argo-20230720-48394-1xggx70/ui/node_modules/webpack/lib/NormalModule.js:417:16)
      at handleParseError (/private/tmp/argo-20230720-48394-1xggx70/ui/node_modules/webpack/lib/NormalModule.js:471:10)
      at /private/tmp/argo-20230720-48394-1xggx70/ui/node_modules/webpack/lib/NormalModule.js:503:5
      at /private/tmp/argo-20230720-48394-1xggx70/ui/node_modules/webpack/lib/NormalModule.js:358:12
      at /private/tmp/argo-20230720-48394-1xggx70/ui/node_modules/loader-runner/lib/LoaderRunner.js:373:3
      at iterateNormalLoaders (/private/tmp/argo-20230720-48394-1xggx70/ui/node_modules/loader-runner/lib/LoaderRunner.js:214:10)
      at Array.<anonymous> (/private/tmp/argo-20230720-48394-1xggx70/ui/node_modules/loader-runner/lib/LoaderRunner.js:205:4)
      at Storage.finished (/private/tmp/argo-20230720-48394-1xggx70/ui/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:55:16)
      at /private/tmp/argo-20230720-48394-1xggx70/ui/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:91:9
      at /private/tmp/argo-20230720-48394-1xggx70/ui/node_modules/graceful-fs/graceful-fs.js:123:16
      at FSReqCallback.readFileAfterClose [as oncomplete] (node:internal/fs/read_file_context:68:3) {
    opensslErrorStack: [ 'error:03000086:digital envelope routines::initialization error' ],
    library: 'digital envelope routines',
    reason: 'unsupported',
    code: 'ERR_OSSL_EVP_UNSUPPORTED'
  }
  
  Node.js v20.4.0
  error Command failed with exit code 1.
  info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
  make: *** [ui/dist/app/index.html] Error 1
```

relates to https://github.com/Homebrew/homebrew-core/pull/137118

### Version

3.4.9

### PR DESCRIPTION
**Pull request by chenrui333**
👋 node 16 is gonna be  EOL on 11 Sep 2023, right now, building with node v20 would have [`digital envelope routines::unsupported` issue](https://github.com/argoproj/argo-workflows/issues/11404), thus updating the build to use v20.

closes #11404
relates to https://github.com/Homebrew/homebrew-core/pull/137118

cc @templarfelix @JPZ13

### CODE REVIEW

**Comment by terrytangyuan** [comment]
I remember you had a comment on the node version, right? @jmeridth

**Comment by terrytangyuan** [comment]
@tczhao @JPZ13 Could you help review this?

**Comment by jmeridth** [comment]
@terrytangyuan was regarding https://github.com/argoproj/argo-workflows/issues/10126 

my issue was that I was able to get a feature working with v18 and then we (argo-workflows) went back to v16.  Feature broke. Going to v20 would be awesome.

**Comment by terrytangyuan** [comment]
Thanks.

Please fix the CI build failures.

File: ui/package.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
         "src"
     ],
     "scripts": {
-        "build": "rm -Rf dist && NODE_OPTIONS='' NODE_ENV=production webpack --mode production --config ./src/app/webpack.config.js",
+        "build": "rm -Rf dist && NODE_OPTIONS='--openssl-legacy-provider' NODE_ENV=production webpack --mode production --config ./src/app/webpack.config.js",
         "start": "NODE_OPTIONS='--no-experimental-fetch' webpack-dev-server --config ./src/app/webpack.config.js",
         "lint": "tslint --fix -p ./src/app",
         "test": "jest"
```
Comment (agilgur5) [inline ui/package.json:12]: on line 15 below, `engines.node` should be updated as well

**Review by agilgur5 — COMMENTED** [review]
A few other places we should update too, like the Nixfiles in `dev/nix/flake.nix`.
**EDIT** actually in `flake.nix` not `node-env.nix`, my bad. See below comments

File: ui/package.json (hunk 1 above)
Comment (chenrui333) [reply ui/package.json:12]: let me add that.

**Comment by chenrui333** [comment]
> A few other places we should update too, like the Nixfiles in `dev/nix/node-env.nix`

do you have any guide about it?

**Comment by agilgur5** [comment]
> do you have any guide about it?

My bad, it's actually in [`flake.env`](https://github.com/argoproj/argo-workflows/blob/master/dev/nix/flake.nix). My Nix is _very_ rusty but I can pull up something tomorrow probably.

If you're not sure, I'm ok approving without so long as CI passes. This is a good change to have and we can update Nix in a follow-up PR 🙂

**Comment by isubasinghe** [comment]
@agilgur5 I can push the Nix stuff, I was the one who ported the build process over to Nix. 
Should be a simple 5 min job on my end.

**Comment by alexec** [comment]
Build error is this: "Error: error:0308010C:digital envelope routines::unsupported"

**Comment by alexec** [comment]
image

File: ui/package.json [hunk 2]
```diff
@@ -6,13 +6,13 @@
         "src"
     ],
     "scripts": {
-        "build": "rm -Rf dist && NODE_OPTIONS='' NODE_ENV=production webpack --mode production --config ./src/app/webpack.config.js",
+        "build": "rm -Rf dist && NODE_OPTIONS='--openssl-legacy-provider' NODE_ENV=production webpack --mode production --config ./src/app/webpack.config.js",
         "start": "NODE_OPTIONS='--no-experimental-fetch' webpack-dev-server --config ./src/app/webpack.config.js",
```
Comment (alexec) [inline ui/package.json:10]:
```suggestion
        "start": "NODE_OPTIONS='--no-experimental-fetch --openssl-legacy-provider' webpack-dev-server --config ./src/app/webpack.config.js",
```

**Comment by alexec** [comment]
@isubasinghe have you considered presenting on Nix to community meeting? Nix is new hotness.

File: ui/package.json (hunk 2 above)
Comment (alexec) [reply ui/package.json:10]: must fix anywhere with `NODE_OPTIONS`

**Comment by isubasinghe** [comment]
@alexec unfortunately my timezone makes attending the community meetings near impossible :(

**Review by terrytangyuan — APPROVED** [review]

**Comment by terrytangyuan** [comment]
https://github.com/argoproj/argo-workflows/actions/runs/5676115276/job/15382363976

```
error argo-ui@1.0.0: The engine "node" is incompatible with this module. Expected version ">=20". Got "18.17.0"
```

**Comment by terrytangyuan** [comment]
Tracking in https://github.com/argoproj/argo-workflows/issues/11459

**Comment by agilgur5** [comment]
> Nix is new hotness.

[2 decades](https://en.wikipedia.org/wiki/Nix_(package_manager)) for some more mainstream attention 😅 

> @alexec unfortunately my timezone makes attending the community meetings near impossible :(

I have yet to go to one but I imagine you could pre-record something which could just be played during the meeting?
Also regarding NPM, did you use [`node2nix`](https://github.com/svanderburg/node2nix) to generate the Nix expressions? It looks vaguely similar, but [docs](https://argoproj.github.io/argo-workflows/running-nix/) don't mention it. Last time I used Nix, I think NPM didn't exist yet 😅

**Comment by isubasinghe** [comment]
@agilgur5 I did use node2nix, its a bit of a manual process to use it with flakes however. 
Probably not the best actually, I am planning to rewrite it using mkDerivation. 
node2nix was just a quicker way of getting something up. 

Also yeah I think I could probably pre-record something instead, will talk to someone about this later.

**Pull request merged**